dask/tests/test_task_spec.py

```diff
@@ -815,3 +816,91 @@ def test_dependencies_mapping_doesnt_mutate_task():
     # The getitem is doing weird stuff and could mutate the task state
     deps[t2.key]
     assert t2.dependencies == {"key"}
+
+
+def test_fuse_tasks_key():
+    a = Task("key-1", func, "a", "b")
+    b = Task("key-2", func2, a.ref(), "d")
+    for t1, t2 in itertools.permutations((a, b)):
+        fused = Task.fuse(t2, t1)
+        assert fused.key == b.key
+
+        fused = Task.fuse(t2, t1, key="new-key")
+        assert fused.key == "new-key"
+
+
+def test_fuse_tasks():
+    a = Task("key-1", func, "a", "b")
+    b = Task("key-2", func2, a.ref(), "d")
+    for t1, t2 in itertools.permutations((a, b)):
+        fused = Task.fuse(t2, t1)
+
+        assert fused() == func2(func("a", "b"), "d")
+
+        t1 = Task("key-1", func, TaskRef("dependency"), "b")
+        t2 = Task("key-2", func2, t1.ref(), "d")
+
+        fused = Task.fuse(t2, t1)
+        assert fused.dependencies == {"dependency"}
+
+        assert fused({"dependency": "dep"}) == func2(func("dep", "b"), "d")
```

Comment:
```suggestion
def test_fuse_tasks():
    a = Task("key-1", func, "a", "b")
    b = Task("key-2", func2, a.ref(), "d")
    c = Task("key-3", func3, b.ref(), "e")
    for t1, t2, t3 in itertools.permutations((a, b, c)):
        fused = Task.fuse(t3, t2, t1)

        assert fused() == func3(func2(func("a", "b"), "d"), "e")

        t1 = Task("key-1", func, TaskRef("dependency"), "b")
        t2 = Task("key-2", func2, t1.ref(), "d")
        t3 = Task("key-3", func3, t2.ref(), "e")
        fused = Task.fuse(t3, t2, t1)
        assert fused.dependencies == {"dependency"}

        assert fused({"dependency": "dep"}) == func3(func2(func("dep", "b"), "d"), "e")
```

It took me a moment to understand that fuse is indeed order-independent, even in the case of transitive dependencies. Extending this test may or may not help. Feel free to ignore.